### ISSUE TITLE
CA1065 should not be raised for ArgumentNullException in IEqualityComparer<T>.GetHashCode

### ISSUE THREAD

**Issue by dfederm** [dotnet/roslyn-analyzers#2994]
#### Analyzer package
Microsoft.CodeAnalysis.FxCopAnalyzers

#### Package Version
v2.9.2

#### Diagnostic ID
CA1065

#### Repro steps
```
    public class SomeEqualityComparer : IEqualityComparer<string>
    {
        public bool Equals(string x, string y)
        {
            return true;
        }

        public int GetHashCode(string obj)
        {
            if (obj == null)
            {
                // This raises CA1065!
                throw new ArgumentNullException(nameof(obj));
            }

            return 0;
        }
    }
```

#### Expected behavior
No CA violation raised.

The documented expectation for `IEqualityComparer<T>.GetHashCode` is to throw null when given a null object. See: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode?redirectedfrom=MSDN&view=netframework-4.8#exceptions

#### Actual behavior
CA1065 is raised.

**Comment by sharwell** [comment]
Even though the documentation allows for this exception, I do not agree that it is expected. I would only be in favor of this change if CA1065 continued to be reported for `GetHashCode` methods implementing `IEqualityComparer<T>` for a nullable type `T`. In other words, the exception would _only_ be made for the case where _both_ of these conditions are met:

1. Nullable reference types is enabled
2. The type implements `IEqualityComparer<T>` for a non-nullable reference type `T`

**Comment by dfederm** [comment]
@sharwell Are you implying the documentation is wrong then? The "correct" way to implement `IEqualityComparer<T>.GetHashCode` would be to throw on `null` (assuming `T` is a reference type).

Regarding your suggestion, I don't think it makes sense. You're suggesting that an exception only be allowed to throw if nullable reference types is enabled, and then `T` is non-nullable, but wouldn't that be enforced by the compiler regardless?

**Comment by sharwell** [comment]
> Are you implying the documentation is wrong then?

I am saying the documentation is misleading compared to what I would consider a best practice. Equality comparers are more versatile if they allow null for `GetHashCode`, similar to how `EqualityComparer<T>.Default` allows null. However, if the equality comparer explicitly implements a type that does not allow null, then it would be fine to throw this exception if null was passed to `GetHashCode`.

> You're suggesting that an exception only be allowed to throw if nullable reference types is enabled, and then T is non-nullable, but wouldn't that be enforced by the compiler regardless?

The compiler helps avoid misuse of null, but does not guarantee it. Exposed APIs should still validate arguments to ensure that the caller did not incorrectly pass null.

**Comment by dfederm** [comment]
What about the docs here: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1065?view=vs-2019#gethashcode-methods

> The versions of GetHashCode that take an argument can throw an ArgumentException. However, Object.GetHashCode should never throw an exception.

`ArgumentNullException` derives from `ArgumentException`, so this should be allowed, right?

**Comment by dfederm** [comment]
Relevant code: https://github.com/dotnet/roslyn-analyzers/blob/7ad9f725f22c7f5a9e8f70a5ebe6e5fd4c14d8d6/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotRaiseExceptionsInUnexpectedLocations.cs#L186-L188

Seems like there's a bug here since it looks like it's trying to explicitly allow `ArgumentException`, but for whatever reason the `IsAssignableTo` check isn't working as expected.

https://github.com/dotnet/roslyn-analyzers/blob/7ad9f725f22c7f5a9e8f70a5ebe6e5fd4c14d8d6/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotRaiseExceptionsInUnexpectedLocations.cs#L101-L106

**Comment by sharwell** [comment]
> Seems like there's a bug here since it looks like it's trying to explicitly allow ArgumentException, but for whatever reason the IsAssignableTo check isn't working as expected.

Interesting that it's not working. It does seem to be coded to allow the exception.

**Comment by Evangelink** [comment]
@dfederm I have just added a unit test with your sample code and I cannot reproduce the issue. There is no diagnostic raised. Maybe something was fixed between the version you are quoting and the latest but I'd like to make sure that this is not linked to something missing in the repro case.

**Comment by dfederm** [comment]
@Evangelink will send you the actual repro offline.

Note that we are using v2.9.2, which might be relevant (we haven't upgraded since upgrading brings in a whole bunch of new rules and thus violations).

**Comment by Evangelink** [comment]
@dfederm Yeah sorry for I would need a not Microsoft flag on my pic ;) 

Maybe @mavasani or @sharwell could have a look at your code if they have some time.

**Comment by EamonNerbonne** [comment]
I think CA1065 is valid here.  Yes, the documentation allows for ArgumentNullException.  However: it does not for `Equals`, and if an implementation does not throw for Equals, and Equals(null, null) hold, then the invariant states that GetHashCode should return the same value.  Throwing an exception in GetHashCode for values permitted by Equals makes using equality comparer's pretty tricky; all of the sudden it matters whether you call GetHashCode or not on a value you may be checking for equality.

Also, existing mscorlib equality comparer's I can find do not throw on null.

In short: as a code-quality issue, for least surprises, GetHashCode should not throw unless Equals would too, and that's ideally never.  The docs may permit it; but it's never necessary, and ideally *should* be avoided: a valid CA1065 therefore.

**Comment by Evangelink** [comment]
@mavasani How do you want to move forward on this issue?

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by sharwell** [comment]
It appears that CA1065 is implemented to explicitly allow `System.ArgumentException` within an implementation of `IEqualityComparer<T>.GetHashCode`:

https://github.com/dotnet/roslyn-analyzers/blob/690be1702a6ae8f915da3bf6cc1b935e3ff264dd/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotRaiseExceptionsInUnexpectedLocations.cs#L185-L187

However, there may be a bug in the implementation of `IsGetHashCodeInterfaceImplementation`:

https://github.com/dotnet/roslyn-analyzers/blob/690be1702a6ae8f915da3bf6cc1b935e3ff264dd/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotRaiseExceptionsInUnexpectedLocations.cs#L271-L302

**Comment by Evangelink** [comment]
@sharwell On PR #3328, I have explicitly added a unit-test for the case mentioned before and there is no diagnostic reported. Until I have some reproducible code, I don't know how to fix the issue.
